pkg/ring/replication_set.go

```diff
@@ -40,25 +49,38 @@ func (r ReplicationSet) Do(ctx context.Context, delay time.Duration, f func(cont
 			}
 			result, err := f(ctx, ing)
 			if err != nil {
-				errs <- err
+				errs <- instanceError{
+					err:      err,
+					instance: ing,
+				}
 			} else {
 				resultsChan <- result
 			}
 		}(i, &r.Ingesters[i])
 	}
 
 	var (
-		numErrs    int
-		numSuccess int
-		results    = make([]interface{}, 0, len(r.Ingesters))
+		numErrs          int
+		numSuccess       int
+		results          = make([]interface{}, 0, len(r.Ingesters))
+		zoneFailureCount = make(map[string]int)
```

Comment:
Number of failures within the zone is not used. To avoid confusion about why using `int` as value type, we can use `map[string]struct{}` instead.

Reply:
Good idea. Originally I thought the count for each zone was needed but that was not the case.